homeassistant/components/sunsynk/config_flow.py

```diff
@@ -0,0 +1,159 @@
+        """Test if we can authenticate with the host."""
+        try:
+            if self.client is None:
+                self.client = await SunsynkClient.create(username, password)
+            else:
+                self.client.username = username
+                self.client.password = password
+                self.client = await self.client.login()
+            return True
+        except InvalidCredentialsException:
+            return False
+
+    async def get_inverters(self) -> list[Inverter]:
+        """Get the list of inverters."""
+        if self.client is None:
+            return []
+        return await self.client.get_inverters()
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> SunsynkHub:
+    """Validate the user input allows us to connect.
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+
+    if CONF_USERNAME not in data or CONF_PASSWORD not in data:
+        raise InvalidAuth
+
+    hub = SunsynkHub()
+
+    if not await hub.authenticate(data[CONF_USERNAME], data[CONF_PASSWORD]):
+        raise InvalidAuth
+
+    return hub
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Sunsynk."""
+
+    VERSION = 1
+
+    username: str | None = None
+    password: str | None = None
+    hub: SunsynkHub | None = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        errors: dict[str, str] = {}
+
+        if user_input is not None:
+            try:
+                self.hub = await validate_input(self.hass, user_input)
+                self.username = user_input[CONF_USERNAME]
+                self.password = user_input[CONF_PASSWORD]
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidAuth:
+                errors["base"] = "invalid_auth"
+            except Exception:  # pylint: disable=broad-except
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+            else:
+                return await self.async_step_inverter()
+
+        return self.async_show_form(
+            step_id=STEP_USER, data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+        )
+
+    async def async_step_inverter(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the inverter selection step."""
+        if user_input is None and self.hub is not None:
+            return self.async_show_form(
+                step_id=STEP_INVERTER, data_schema=await self._inverter_scheme()
+            )
+
+        assert user_input is not None
+
+        if self._inverter_already_exists(user_input[DATA_INVERTER_SN]):
+            raise AbortFlow("already_configured")
+
+        await self.async_set_unique_id(user_input[DATA_INVERTER_SN])
```

Comment:
If you first set the unique_id, and then `self._abort_if_unique_id_configured()`, will do exactly what you want

Reply:
Thanks, done